File: utils/mpmt/mpmt1.scala
/**
 * mpmt1.scala: A Scala version of mpmt1
 *
 * STATUS:
 *  Under development
 * License:
 *   Apache License, Version 2.0
 * History:
 *   2022/01/01 v0.1 Initial version.
 * Author:
 *   Masanori Itoh <masanori.itoh@gmail.com>
 * TODO:
 *   * Write multi process version?
 */

//break requires scala 2.8 or later.
//import scala.util.control.Breaks._

object mpmt1 {
  def main(args: Array[String]) = {
    var num_context = 4
    var duration = 30
    var mode = "t"

    //Parse options.
    //NOTE(thatsdone): Avoided to use getopt pacakge for now.
    for (i <- 0 until args.length) {
      if (args(i) == "-n") {
        if (args(i + 1).startsWith("-")) {
          println("-n requires a value")
          sys.exit()
        }
        num_context = args(i + 1).toInt

      } else if (args(i) == "-d") {
        if (args(i + 1).startsWith("-")) {
          println("-d requires a value")
          sys.exit()
        }
        duration = args(i + 1).toInt
      } else if (args(i) == "-m") {
        //ignore at the moment
        if (args(i + 1).startsWith("-")) {
          println("-m requires a value")
          sys.exit()
        }
        //mode = args(i + 1)
        println("Currently -m option is ignored.")
      } else {
        //ignore at the moment
      }
    }

    printf("num_context: %d duration %d (s) mode: %s\n",
           num_context, duration, mode)

    val threads = {
      for (i <- 0 until num_context)
        yield new Thread(new BusyWorker(num_context, duration))
    }
    threads.foreach({
      t => t.start()
      println("Thread started: " + t)
    })
    threads.foreach({
      t => t.join()
      println("Thread completed: " + t)
    })
  }
}

class BusyWorker(num_context: Int, duration: Int) extends Runnable {

  def run() = {
    var ts_save = System.currentTimeMillis()
    var ts = ts_save
    //duration is in seconds. max is in milli-seconds
    val max = duration * 1000

    //Don't use Scala3 syntax for now.
    while ({
      ts = System.currentTimeMillis()
      (ts - ts_save) <= max
    }) ()
    //BTW, Scala does not have 'break'
    println("Expired! " + (ts - ts_save))
  }
}

File: utils/mpmt/mpmt1.js
/*
# -*- coding: utf-8 -*-
#
# mpmt1.rb: A Node.js version of mpmt1.py
#
# License:
#   Apache License, Version 2.0
# History:
#   * 2021/12/30 v0.1 Initial version
# Author:
#   Masanori Itoh <masanori.itoh@gmail.com>
# TODO:
#   * Investigate multi-thread feature of Node.js
#   * Add options by using getopt
*/

function busy_worker(timeout) {

    // ts and ts_save are in mili-seconds
    var ts_save = Date.now()
    // duration is in seconds.
    var max = timeout * 1000

    while (true) {
	ts = Date.now()
	if ((ts - ts_save) >= max) {
	    console.log('Expired! ' + (ts - ts_save))
	    break
	}
    }
    process.exit(123)
}


var posix_getopt = require('posix-getopt')
var parser, option

const cluster = require('cluster')

var num_context = 4
var duration = 10
var mode = 'p'

parser = new posix_getopt.BasicParser('n:d:m', process.argv)
while ((option = parser.getopt()) !== undefined) {
    switch (option.option) {
    case 'n':
	num_context = option.optarg
	break;
    case 'd':
	duration = option.optarg
	break;
    case 'm':
	mode = option.optarg
	break;
    }
}


//isPrimary does not work for Node.js v10.19.0?
if (cluster.isMaster) {
    console.log('parent: pid= ' + process.pid)
    console.log('num_context: ' + num_context + ' duration: ' + duration)
    for (var i = 0; i < num_context; i++) {
	cluster.fork()
    }

    cluster.on('exit', (code, signal) => {
	// signal is the value for process.exit()? node code?
	console.log('on exit called. code= ' + code + ' signal= ' + signal)
    });

} else {
    console.log('child: pid=' + process.pid)
    busy_worker(duration)
}

